### PR TITLE
Add `distinct()` API to repeater

### LINKED ISSUE #7109: `distinct` validation rule

### PR DESCRIPTION
**Pull request by cheesegrits**
- [x] Changes have been thoroughly tested to not break existing functionality.
- [x] New functionality has been documented or existing documentation has been updated to reflect changes.
- [ ] Visual changes are explained in the PR description using a screenshot/recording of before and after.

PR for distinct() functionality in Repeaters, as discussed in Discord.

I over-commented the main trait, just to make the design decisions clear and explain some of the gnarlier collect() code, can of course strip that out.

The code feels a little scrappy, so I'm totally open to suggestions for improving it.  But it all works.

Lmk if you want me to add any tests for any of this.

I'm not sure if I used the right syntax for inter-document links (in the Repeater doc).  Wherever you guys do that, you use links like 'select#slugified-section-name', but I could only get it to work with the full document name, like '05-radio.md#slugified-section-name'.

Each field in action:

https://app.screencast.com/U9Fn53PUZ6V9I

### CODE REVIEW

**Comment by zepfietje** [comment]
Maybe I incorrectly linked #7109 and confused it with our plans for a `distinct()` validation rule?

Anyway, this API would collide with those plans then.
I do think both features try to achieve the same thing, one using validation and another by updating other repeater items.
From a UX perspective I do want to note this may be confusing to users if fields automatically update, don't you think?

**Comment by danharrin** [comment]
@zepfietje I've discussed this with Hugh already. It will be paired with validation. It's deliberate that other fields in the repeater update. For example, imagine if you have a Question resource with an `answers` repeater, and only 1 answer can be marked as correct at a time

**Comment by zepfietje** [comment]
Yeah I do like the functionality, and it makes sense in the case you described.
Maybe adding a hint to the field suffices to clarify when the UX is a little different and it's unclear that other repeater items will be updated.

So fine by me then. 👍

**Comment by zepfietje** [comment]
By the way, shouldn't other fields like text input also get support for this API, if even just the validation part?

**Comment by danharrin** [comment]
Yeah, probably just the validation for the text input / color picker.

**Comment by cheesegrits** [comment]
I was wondering when implementing this whether the validation should be separated, so it could be used independently, just automatically used when doing distinct selection.  So maybe a distinctSelection() method, which implicitly also uses a distinct() validation method, which itself could be used independently.

**Comment by zepfietje** [comment]
Yeah I think that might be best actually.

**Comment by cheesegrits** [comment]
@danharrin what do you reckon about splitting the validation out, so there's two methods ...

distinctSelection() - handles the selection / de-selection / disabling logic
distinct() - just validation

... and have distinctSelection() automatically apply the distinct() validation.  Then distinct() validation could be applied to columns that don't support the selection logic.  Or indeed, applied to Select etc, but only validate, don't police the actual selection.

**Comment by cheesegrits** [comment]
As per discussion in Discord DM with Dan, I split out distinct() validation from the (renamed) distinctOptions() method.  If you specify distinctOptions() it will automatically add the distinct() validation, but distinct() validation can be used without distinctOptions().

Ooops, forgot to change the documentation.  Will commit that soon.

**Comment by danharrin** [comment]
Can we lean on the existing `validation.distinct` translation string from Laravel?

**Comment by zepfietje** [comment]
> Can we lean on the existing `validation.distinct` translation string from Laravel?

I think we can, yeah.
Won't suit all cases, but users can always write their own message if they want something more specific.

**Comment by danharrin** [comment]
There are 3 translation strings at the moment, I think we can at least replace one of them.

**Comment by cheesegrits** [comment]
Btw, do you prefer distinctSelection() or distinctOptions() as the method name? Or somethiny else?

File: packages/forms/docs/03-fields/03-select.md [hunk 1]
```diff
@@ -407,6 +407,48 @@ Select::make('status')
...
+
+You may enforce selection of any given option in only one instance of a Repeater by using the `distinctOptions()` method.
+
+This example of creating teams for a tournament in a Repeater would only allow each shirt color to be chosen once:
+
+```php
+use Filament\Forms\Components\Repeaater;
+use Filament\Forms\Components\Select;
+
+Repeaater::make('teams')
```
Comment (valpuia604) [inline packages/forms/docs/03-fields/03-select.md]: Typo for Repeater

File: packages/forms/docs/03-fields/04-checkbox.md [hunk 2]
```diff
@@ -54,6 +54,28 @@ Checkbox::make('is_admin')->inline(false)
...
+
+You may enforce selection of a checkbox in one (and only one) instance of a Repeater by using the `distinctOptions()` method.
+
+This example of adding multiple locations in a Repeater would require that one (and only one) instance is selected as primary: 
+
+```php
+use Filament\Forms\Components\Checkbox;
+use Filament\Forms\Components\Repeaater;
+
+Repeaater::make('locations')
```
Comment (valpuia604) [inline packages/forms/docs/03-fields/04-checkbox.md]: Here also

File: packages/forms/docs/03-fields/05-toggle.md [hunk 3]
```diff
@@ -86,6 +86,28 @@ Toggle::make('is_admin')
...
+
+You may enforce selection of a Toggle in one (and only one) instance of a Repeater by using the `distinctOptions()` method.
+
+This example of adding multiple locations in a Repeater would require that one (and only one) instance is selected as primary:
+
+```php
+use Filament\Forms\Components\Checkbox;
+use Filament\Forms\Components\Repeaater;
+
+Repeaater::make('locations')
```
Comment (valpuia604) [inline packages/forms/docs/03-fields/05-toggle.md]: again :rofl:

File: packages/forms/docs/03-fields/06-checkbox-list.md [hunk 4]
```diff
@@ -132,6 +132,36 @@ CheckboxList::make('technologies')
...
+
+You may enforce selection of any given checkbox in only one instance of a Repeater by using the `distinctOptions()` method.
+
+This example of assigning pets to rooms in your house would only allow each pet to be assigned to one room:
+
+```php
+use Filament\Forms\Components\CheckboxList;
+use Filament\Forms\Components\Repeaater;
+
+Repeaater::make('rooms')
```
Comment (valpuia604) [inline packages/forms/docs/03-fields/06-checkbox-list.md]: ahhh

File: packages/forms/docs/03-fields/07-radio.md [hunk 5]
```diff
@@ -101,3 +101,33 @@ Radio::make('status')
...
+
+You may enforce selection of any given radio button in only one instance of a Repeater by using the `distinctOptions()` method.
+
+This example of creating teams for a tournament in a Repeater would only allow each shirt color to be chosen once
+
+```php
+use Filament\Forms\Components\Radio;
+use Filament\Forms\Components\Repeaater;
+
+Repeaater::make('teams')
```
Comment (valpuia604) [inline packages/forms/docs/03-fields/07-radio.md]: last

**Comment by danharrin** [comment]
Thank you very much! This is very cool.

**Pull request merged**